services/web/server/src/simcore_service_webserver/projects/projects_api.py

```diff
@@ -253,6 +271,86 @@ async def _get_default_pricing_and_hardware_info(
     )
 
 
+_RAM_SAFE_MARGIN_RATIO: Final[
+    float
+] = 0.1  # NOTE: machines always have less available RAM than advertised
+_CPUS_SAFE_MARGIN: Final[float] = 0.1
+
+
+async def update_project_node_resources_from_hardware_info(
+    app: web.Application,
+    *,
+    user_id: UserID,
+    project_id: ProjectID,
+    node_id: NodeID,
+    product_name: str,
+    hardware_info: HardwareInfo,
+) -> None:
+    if not hardware_info.aws_ec2_instances:
+        return
+    try:
+        rabbitmq_rpc_client = get_rabbitmq_rpc_client(app)
+        unordered_list_ec2_instance_types: list[
+            EC2InstanceTypeGet
+        ] = await get_instance_type_details(
+            rabbitmq_rpc_client,
+            instance_type_names=set(hardware_info.aws_ec2_instances),
+        )
+
+        assert unordered_list_ec2_instance_types  # nosec
+
+        # NOTE: with the current implementation, there is no use to get the instance past the first one
+        def _by_type_name(ec2: EC2InstanceTypeGet) -> bool:
+            return bool(ec2.name == hardware_info.aws_ec2_instances[0])
+
+        selected_ec2_instance_type = next(
+            iter(filter(_by_type_name, unordered_list_ec2_instance_types))
+        )
+
+        # now update the project node required resources
+        # NOTE: we keep a safe margin with the RAM as the dask-sidecar "sees"
+        # less memory than the machine theoretical amount
+        db: ProjectDBAPI = ProjectDBAPI.get_from_app_context(app)
+        node = await db.get_project_node(project_id, node_id)
+        node_resources = parse_obj_as(ServiceResourcesDict, node.required_resources)
+        if DEFAULT_SINGLE_SERVICE_NAME in node_resources:
+            image_resources: ImageResources = node_resources[
+                DEFAULT_SINGLE_SERVICE_NAME
+            ]
+            image_resources.resources["CPU"].set_value(
+                float(selected_ec2_instance_type.cpus) - _CPUS_SAFE_MARGIN
+            )
+            image_resources.resources["RAM"].set_value(
+                int(
+                    selected_ec2_instance_type.ram
+                    - _RAM_SAFE_MARGIN_RATIO * selected_ec2_instance_type.ram
```

Comment:
you could also express it as `selected_ec2_instance_type.ram * 0.9` where 0.9 can be called `_SAFE_TOTAL_RAM_PERCENT`?

Reply:
a percent goes from 1 - 100
a ratio from 0 - 1